terra_notebook_utils/progress.py

```diff
@@ -0,0 +1,126 @@
+import sys
+import time
+import threading
+import sched
+from math import floor, ceil
+from contextlib import AbstractContextManager
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+
+class RateLimited:
+    """
+    Decorator to rate limit method calls. Raises `exception` if method is called more quickly than
+    `rate` times per second, or does nothing if exception is None.
+    """
+    def __init__(self, rate: float, exception: Exception=None):
+        self.period = 1 / rate
+        self.exception = exception
+        self.reset()
+
+    def reset(self):
+        self._last_call = time.time() - 10 * self.period
```

Comment:
What does the `10` account for?

Reply:
It's just an arbitrary factor to guarantee the function is not rate limited on the next call. anything larger than 1 should work.